### PR TITLE
[SPARK-29107][SQL][TESTS] Port window.sql (Part 1)

### PR DESCRIPTION
**Pull request by DylanGuedes**
### What changes were proposed in this pull request?

This PR ports window.sql from PostgreSQL regression tests https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/sql/window.sql from lines 1~319

The expected results can be found in the link: https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/expected/window.out

## How was this patch tested?
Pass the Jenkins.

### Why are the changes needed?
To ensure compatibility with PGSQL

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Comparison with PgSQL results.

### CODE REVIEW

**Comment by wangyum** [comment]
ok to test

**Comment by shaneknapp** [comment]
test this please

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 1]
```diff
@@ -0,0 +1,343 @@
+-- Portions Copyright (c) 1996-2019, PostgreSQL Global Development Group
+--
+-- Window Functions Testing
+-- https://github.com/postgres/postgres/blob/REL_12_BETA3/src/test/regress/sql/window.sql
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: Can you add line numbers?: https://github.com/apache/spark/blob/a6a663c4379390217443bc5b6f75873fb1c38c73/sql/core/src/test/resources/sql-tests/inputs/pgSQL/aggregates_part1.sql#L6

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+-- FROM tenk1 s WHERE unique2 < 10;
+
+-- empty table
+SELECT count(*) OVER (PARTITION BY four) FROM (SELECT * FROM tenk1 WHERE FALSE)s;
+
+-- mixture of agg/wfunc in the same window
+SELECT sum(salary) OVER w, rank() OVER w FROM empsalary WINDOW w AS (PARTITION BY depname ORDER BY salary DESC);
+
+-- strict aggs
+-- Temporarily turns off the ANSI mode because of compatibility issues between keywords
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: Can you describe which keyword?

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+SELECT sum(unique1) over (rows between 1 following and 3 following),
+unique1, four
+FROM tenk1 WHERE unique1 < 10;
+
+SELECT sum(unique1) over (rows between unbounded preceding and 1 following),
+unique1, four
+FROM tenk1 WHERE unique1 < 10;
+
+-- [SPARK-28428] Spark `exclude` always expecting `()`
+SELECT sum(unique1) over (w range between current row and unbounded following),
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: Why didn't you comment out this statement? You did the other statements, though?

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+--   exclude no others) as sum_rows FROM generate_series(1, 10) i;
+-- SELECT * FROM v_window;
+
+-- [SPARK-28648] Adds support to `groups` unit type in window clauses
+-- CREATE OR REPLACE TEMP VIEW v_window AS
+-- SELECT i.id, sum(i.id) over (order by i.id groups between 1 preceding and 1 following) as sum_rows FROM range(1, 11) i;
+-- SELECT * FROM v_window;
+
+DROP VIEW v_window;
+DROP TABLE empsalary;
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: How about `tenk2` and `int4_tbl`?

**Review by maropu — APPROVED** [review]
LGTM except for some minor comments. @wangyum @HyukjinKwon

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: Unfortunately,`REL_12_BETA4` is released.
> https://github.com/postgres/postgres/blob/REL_12_BETA4/src/test/regress/sql/window.sql

Comment (DylanGuedes) [reply sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: I can update to match Beta4, but:
- If the PgSQL version is *that* relevant, then we will add a new JIRA for every new PgSQL release for every .sql migration (i.e:  strings.sql, union.sql, date.sql, aggregates.sql, etc). But we didn't.
- If the PgSQL is not that relevant, then there's no need to change PRs to match the latest release candidate version (although it matches a stable version).

What you think?

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: Have you checked the last discussion? https://github.com/apache/spark/pull/24850#discussion_r300215383
I think we don't need to stick to the single snapshot, so its ok to update it to `REL_12_BETA4` in this port. Actually, no difference between `BETA2` and `BETA4` in `window.sql`?

**Comment by HyukjinKwon** [comment]
ping @DylanGuedes to update

**Comment by DylanGuedes** [comment]
@HyukjinKwon I updated with the changes suggested, but still not clear: should I update to REL4?

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql (hunk 1 above)
Comment (DylanGuedes) [reply sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: No, I didn't get noticed that you guys were already discussing about updating other tests.
I checked last week and there are some major differences between BETA2 and BETA4 for `window.sql`, I just can't remember if lines L1-L319 are impacted by them. So, it's ok to keep at REL2 or should I updated to REL3?

**Comment by HyukjinKwon** [comment]
Yea, we should update it.

File: sql/core/src/test/resources/sql-tests/results/pgSQL/window_part1.sql.out [hunk 5]
```diff
@@ -0,0 +1,962 @@
...
+2	6	20	8
+2	8	20	8
+3	1	25	9
+3	3	25	9
+3	5	25	9
+3	7	25	9
+3	9	25	9
+
+
+-- !query 43
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/pgSQL/window_part1.sql.out]: This test seems failing when it's ran via the thrift server. Can you check the cause and exclude this in the test? cc @wangyum

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 6]
```diff
@@ -187,10 +187,11 @@ sum(ten) over (partition by four order by ten range between unbounded preceding
...
 
-SELECT four, ten/4 as two,
-sum(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row),
-last(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row)
-FROM (select distinct ten, four from tenk1) ss;
+-- Failing on thrift server
+-- SELECT four, ten/4 as two,
+-- sum(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row),
+-- last(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row)
+-- FROM (select distinct ten, four from tenk1) ss;
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: Ur, did you regenerate the golden files after this removal?

**Comment by dongjoon-hyun** [comment]
Hmm. Still `Result did not match for query ` for the following. 
```
SELECT four, ten/4 as two,
sum(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row),
last(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row)
FROM (select distinct ten, four from tenk1) ss
```

@DylanGuedes . Could you post the query result in both `spark-shell` and `spark-sql`?
Previously, there was a bug which the result is different in both shells and `spark-sql` is similar with Spark Thrift Server.

**Comment by DylanGuedes** [comment]
It works fine locally. And, to be fair, I have no idea why this is not working: in my older PR It also always passed. I'll remote both queries that are failing.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+sum(ten) over (partition by four order by ten range between unbounded preceding and current row),
+last(ten) over (partition by four order by ten range between unbounded preceding and current row)
+FROM (select distinct ten, four from tenk1) ss;
+
+SELECT four, ten,
+sum(ten) over (partition by four order by ten range between unbounded preceding and unbounded following),
+last(ten) over (partition by four order by ten range between unbounded preceding and unbounded following)
+FROM (select distinct ten, four from tenk1) ss;
+
+-- Failing on thrift server
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: We need a JIRA issue if we are not going to cover this query.

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+-- SELECT four, ten/4 as two,
+-- sum(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row),
+-- last(ten/4) over (partition by four order by ten/4 range between unbounded preceding and current row)
+-- FROM (select distinct ten, four from tenk1) ss;
+
+-- Failing on thrift server
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/window_part1.sql]: We need a JIRA issue if we are not going to cover this query, too.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I know this consumed your time your time, @DylanGuedes .

I believe the followings are the final pieces.
- A new JIRA for the newly excluded query 1: https://github.com/apache/spark/pull/25816#discussion_r334258188
- A new JIRA for the newly excluded query 2: https://github.com/apache/spark/pull/25816#discussion_r334258197
- Fix the PR inconsistency. (1) PR description has a link to `REL_12_BETA2`. (2) Code has a link to `REL_12_BETA4`. (3) The latest one is already `REL_12_STABLE`. The best option is to sync with `REL_12_STABLE`. Could you check the diff? I believe there is no diff on this file. Then, we can use `REL_12_STABLE`.

cc @maropu , too.

**Comment by dongjoon-hyun** [comment]
I checked the diff. There is additional addition from line 1257, but that is irrelevant to this PR because this PR is only for line 1 to line 319.
```
~/A/postgres:REL_12_STABLE$ git diff REL_12_BETA4 src/test/regress/sql/window.sql
diff --git a/src/test/regress/sql/window.sql b/src/test/regress/sql/window.sql
index fc6d4cc903..fe273aa31e 100644
--- a/src/test/regress/sql/window.sql
+++ b/src/test/regress/sql/window.sql
@@ -1257,3 +1257,22 @@ SELECT to_char(SUM(n::float8) OVER (ORDER BY i ROWS BETWEEN CURRENT ROW AND 1 FO
 SELECT i, b, bool_and(b) OVER w, bool_or(b) OVER w
   FROM (VALUES (1,true), (2,true), (3,false), (4,false), (5,true)) v(i,b)
   WINDOW w AS (ORDER BY i ROWS BETWEEN CURRENT ROW AND 1 FOLLOWING);
+
+-- Tests for problems with failure to walk or mutate expressions
+-- within window frame clauses.
+
+-- test walker (fails with collation error if expressions are not walked)
+SELECT array_agg(i) OVER w
+  FROM generate_series(1,5) i
+WINDOW w AS (ORDER BY i ROWS BETWEEN (('foo' < 'foobar')::integer) PRECEDING AND CURRENT ROW);
+
+-- test mutator (fails when inlined if expressions are not mutated)
+CREATE FUNCTION pg_temp.f(group_size BIGINT) RETURNS SETOF integer[]
+AS $$
+    SELECT array_agg(s) OVER w
+      FROM generate_series(1,5) s
+    WINDOW w AS (ORDER BY s ROWS BETWEEN CURRENT ROW AND GROUP_SIZE FOLLOWING)
+$$ LANGUAGE SQL STABLE;
+
+EXPLAIN (costs off) SELECT * FROM pg_temp.f(2);
+SELECT * FROM pg_temp.f(2);
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @DylanGuedes , @maropu , @HyukjinKwon !
Merged to master for this.

**Comment by gengliangwang** [comment]
I have created a followup to correct the file location: https://github.com/apache/spark/pull/26116

**Comment by dongjoon-hyun** [comment]
Oops. Yes. We moved the location and this PR is very old.
For now, #26116 is failing. I'll revert this PR. Sorry guys.

**Comment by dongjoon-hyun** [comment]
I missed the location, @DylanGuedes .
Sorry, but could you make a PR after putting the file into the correct location `sql/core/src/test/resources/sql-tests/inputs/postgreSQL` instead of `pgSQL`?
As @gengliangwang pointed out, this PR may fail at that new location. We can decide to file a JIRA or fix it at that time.

And, thank you again, @gengliangwang !

**Comment by DylanGuedes** [comment]
@dongjoon-hyun I [made a new PR](https://github.com/apache/spark/pull/26119) with the new location.

**Pull request closed without merging**